File: packages/shared/src/types/laboratory.ts
/** Lab order priority levels. */
export type LabPriority = "stat" | "urgent" | "routine";

/** Lab order status lifecycle. */
export type LabOrderStatus =
  | "ordered"
  | "collected"
  | "processing"
  | "completed"
  | "cancelled";

/** Lab result status lifecycle. */
export type LabResultStatus = "pending" | "preliminary" | "final" | "cancelled";

/** Interpretation of a lab result. */
export type LabInterpretation =
  | "normal"
  | "abnormal"
  | "critical"
  | "inconclusive";

// ── Lab Worklist ─────────────────────────────────────────────────────────────

/** Lab worklist item (order with patient info and result counts). */
export interface LabWorklistItem {
  id: string;
  order_number: string;
  encounter_id: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  ordered_by: string;
  priority: LabPriority;
  status: LabOrderStatus;
  specimen_type: string | null;
  clinical_info: string | null;
  fasting: boolean;
  test_count: number;
  pending_count: number;
  critical_count: number;
  created_at: string;
}

/** Lab worklist response. */
export interface LabWorklistResponse {
  items: LabWorklistItem[];
  total: number;
}

// ── Lab Order Detail ─────────────────────────────────────────────────────────

/** Lab order response from the API. */
export interface LabOrderDetail {
  id: string;
  encounter_id: string;
  patient_id: string;
  ordered_by: string;
  order_number: string;
  priority: LabPriority;
  status: LabOrderStatus;
  specimen_type: string | null;
  specimen_collected_at: string | null;
  clinical_info: string | null;
  fasting: boolean;
  total_cost_cents: number | null;
  created_at: string;
}

/** Full lab order with results and patient info. */
export interface LabOrderWithResults {
  order: LabOrderDetail;
  results: LabResultDetail[];
  patient_name: string | null;
  patient_mrn: string | null;
}

// ── Lab Results ──────────────────────────────────────────────────────────────

/** Full lab result record. */
export interface LabResultDetail {
  id: string;
  order_id: string;
  patient_id: string;
  performed_by: string | null;
  verified_by: string | null;
  test_code: string;
  test_name: string;
  loinc_code: string | null;
  panel_name: string | null;
  result_value: string | null;
  result_numeric: number | null;
  result_unit: string | null;
  reference_range: string | null;
  interpretation: LabInterpretation | null;
  is_critical: boolean;
  is_abnormal: boolean;
  critical_notified: boolean;
  critical_notified_to: string | null;
  critical_notified_at: string | null;
  status: LabResultStatus;
  resulted_at: string | null;
  verified_at: string | null;
  notes: string | null;
  method: string | null;
  created_at: string;
}

/** Payload for entering a lab result. */
export interface LabResultEntry {
  result_value?: string | null;
  result_numeric?: number | null;
  result_unit?: string | null;
  reference_range?: string | null;
  interpretation?: LabInterpretation | null;
  is_abnormal?: boolean;
  notes?: string | null;
  method?: string | null;
}

/** Payload for verifying a lab result. */
export interface LabResultVerify {
  notes?: string | null;
}

/** Payload for recording critical value notification. */
export interface CriticalNotifyRequest {
  notified_to: string;
}

/** Specimen collection request. */
export interface SpecimenCollectRequest {
  specimen_id?: string | null;
}

File: services/api-gateway/app/routers/laboratory.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import CurrentUser, get_current_user, require_roles
from app.auth.license_check import require_module
from app.database import get_db
from app.schemas.lab import (
    CriticalNotifyRequest,
    LabOrderDetail,
    LabOrderResponse,
    LabResultEntry,
    LabResultResponse,
    LabResultVerify,
    LabWorklistResponse,
    SpecimenCollectRequest,
)
from app.services.lab_service import LabService

router = APIRouter(dependencies=[Depends(require_module("laboratory"))])


# ── Worklist ──────────────────────────────────────────────────────────────────


@router.get("/worklist", response_model=LabWorklistResponse)
async def get_worklist(
    status_filter: str | None = Query(None, alias="status", description="Filter by order status"),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
) -> LabWorklistResponse:
    """
    Get the lab worklist — orders with patient info and result counts.
    Ordered by priority (stat > urgent > routine) then creation time.

    @param status_filter: Optional status filter (ordered|collected|processing|completed)
    @param db: Database session
    @param current_user: Authenticated user from JWT
    @returns Lab worklist with items and total count
    """
    service = LabService(db)
    items, total = await service.get_worklist(
        facility_id=current_user.facility_id,
        status_filter=status_filter,
    )
    return LabWorklistResponse(items=items, total=total)


# ── Order Detail ──────────────────────────────────────────────────────────────


@router.get("/orders/{order_id}", response_model=LabOrderDetail)
async def get_order_detail(
    order_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
) -> LabOrderDetail:
    """
    Get a lab order with all results and patient info.

    @param order_id: LabOrder UUID
    @param db: Database session
    @param current_user: Authenticated user from JWT
    @returns Full lab order detail with results
    """
    service = LabService(db)
    order, results, patient_name, patient_mrn = await service.get_order_detail(
        order_id=order_id,
        facility_id=current_user.facility_id,
    )
    if not order:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lab order not found",
        )
    return LabOrderDetail(
        order=LabOrderResponse.model_validate(order),
        results=[LabResultResponse.model_validate(r) for r in results],
        patient_name=patient_name,
        patient_mrn=patient_mrn,
    )


# ── Specimen Collection ───────────────────────────────────────────────────────


@router.post(
    "/orders/{order_id}/collect",
    response_model=LabOrderResponse,
)
async def collect_specimen(
    order_id: uuid.UUID,
    data: SpecimenCollectRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_roles("lab_tech", "nurse", "admin", "facility_admin")),
) -> LabOrderResponse:
    """
    Record specimen collection for a lab order.

    @param order_id: LabOrder UUID
    @param data: Optional specimen ID / barcode
    @param db: Database session
    @param current_user: Authenticated lab tech or nurse
    @returns Updated lab order
    """
    service = LabService(db)
    order = await service.collect_specimen(
        order_id=order_id,
        facility_id=current_user.facility_id,
        collected_by=current_user.user_id,
        specimen_id=data.specimen_id,
    )
    if not order:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lab order not found",
        )
    return LabOrderResponse.model_validate(order)


# ── Result Entry ──────────────────────────────────────────────────────────────


@router.post(
    "/results/{result_id}/enter",
    response_model=LabResultResponse,
)
async def enter_result(
    result_id: uuid.UUID,
    data: LabResultEntry,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_roles("lab_tech", "admin", "facility_admin")),
) -> LabResultResponse:
    """
    Enter a lab result. Auto-detects critical values.
    Critical lab values MUST trigger immediate alert (CLAUDE.md: Clinical Safety).

    @param result_id: LabResult UUID
    @param data: Result entry data
    @param db: Database session
    @param current_user: Authenticated lab technician
    @returns Updated lab result
    """
    service = LabService(db)
    lab_result = await service.enter_result(
        result_id=result_id,
        data=data,
        facility_id=current_user.facility_id,
        performed_by=current_user.user_id,
    )
    if not lab_result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lab result not found",
        )
    return LabResultResponse.model_validate(lab_result)


# ── Result Verification ──────────────────────────────────────────────────────


@router.post(
    "/results/{result_id}/verify",
    response_model=LabResultResponse,
)
async def verify_result(
    result_id: uuid.UUID,
    data: LabResultVerify,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_roles("lab_tech", "pathologist", "admin", "facility_admin")),
) -> LabResultResponse:
    """
    Verify (approve) a lab result. Changes status from preliminary to final.
    When all results in an order are verified, the order auto-completes.

    @param result_id: LabResult UUID
    @param data: Optional verification notes
    @param db: Database session
    @param current_user: Authenticated senior lab tech or pathologist
    @returns Verified lab result
    """
    service = LabService(db)
    try:
        lab_result = await service.verify_result(
            result_id=result_id,
            facility_id=current_user.facility_id,
            verified_by=current_user.user_id,
            notes=data.notes,
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    if not lab_result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lab result not found",
        )
    return LabResultResponse.model_validate(lab_result)


# ── Critical Value Notification ───────────────────────────────────────────────


@router.post(
    "/results/{result_id}/notify-critical",
    response_model=LabResultResponse,
)
async def notify_critical(
    result_id: uuid.UUID,
    data: CriticalNotifyRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_roles("lab_tech", "admin", "facility_admin")),
) -> LabResultResponse:
    """
    Record that a critical lab value was communicated to a clinician.
    Critical lab values MUST trigger immediate alert (CLAUDE.md: Clinical Safety).

    @param result_id: LabResult UUID
    @param data: Clinician who was notified
    @param db: Database session
    @param current_user: Authenticated lab technician
    @returns Updated lab result with notification recorded
    """
    service = LabService(db)
    lab_result = await service.record_critical_notification(
        result_id=result_id,
        facility_id=current_user.facility_id,
        notified_by=current_user.user_id,
        notified_to=data.notified_to,
    )
    if not lab_result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Lab result not found",
        )
    return LabResultResponse.model_validate(lab_result)
